update_cpes.py

```diff
@@ -263,77 +409,108 @@ def update_cpes(xml_file, cpe_vp_map, r7_vp_map, deprecated_cves):
+
+            if not success:
+                continue
+
+            # Sanity check the value to ensure that no invalid values will
+            # slip in due to logic or mapping bugs.
+            # If it's not in the official NIST list then log it and kick it out
+            if not cpe_vp_map.contains(cpe_fields[0], cpe_fields[1], cpe_fields[2]):
+                logging.error("Invalid CPE type %s created for fields %s. This may be due to an invalid mapping.", cpe_fields[0], cpe_fields[1], cpe_fields[2])
+                continue
+
+            # building the CPE string
+            # Last minute escaping of '/' and `!`
+            cpe_fields = [x.replace("/", "\/").replace("%21", "\!") for x in cpe_fields]
+
+            # Colon separated fields, then the edition field
+            cpe_value = "cpe:/" + ":".join(cpe_fields)
+
+            # Otherwise, remove any trailing wildcard fields
+            cpe_value = re.sub("(:\*)+$", "", cpe_value)
+
+            cpe_param = etree.Element('param')
+            cpe_param.attrib['pos'] = '0'
+            cpe_param.attrib['name'] = '{}.cpe23'.format(fp_type)
+            cpe_param.attrib['value'] = cpe_value
+
+            for param_name in params[fp_type]:
+                param = params[fp_type][param_name]
+                parent = param.getparent()
+                index = parent.index(param) + 1
+                parent.insert(index, cpe_param)
 
     root = doc.getroot()
 
     with open(xml_file, 'wb') as xml_out:
         xml_out.write(etree.tostring(root, pretty_print=True, xml_declaration=True, encoding=doc.docinfo.encoding))
 
-
 def main():
-    if len(sys.argv) != 4:
-        logging.critical("Expecting exactly 3 arguments; recog XML file, CPE 2.3 XML dictionary, JSON remapping, got %s", (len(sys.argv) - 1))
+    if len(sys.argv) != 3:
+        logging.critical("Expecting exactly 2 arguments; CPE 2.3 XML dictionary, YAML remapping, got %s", (len(sys.argv) - 1))
         sys.exit(1)
 
-    cpe_vp_map, deprecated_map = parse_cpe_vp_map(sys.argv[2])
+    cpe_vp_map, deprecated_map = parse_cpe_vp_map(sys.argv[1])
     if not cpe_vp_map:
-        logging.critical("No CPE vendor => product mappings read from CPE 2.3 XML dictionary %s", sys.argv[2])
+        logging.critical("No CPE vendor => product mappings read from CPE 2.3 XML dictionary %s", sys.argv[1])
         sys.exit(1)
 
-    r7_vp_map = parse_r7_remapping(sys.argv[3])
+    r7_vp_map = parse_r7_remapping(sys.argv[2])
     if not r7_vp_map:
-        logging.warning("No Rapid7 vendor/product => CPE mapping read from %s", sys.argv[3])
+        logging.warning("No Rapid7 vendor/product => CPE mapping read from %s", sys.argv[2])
 
     # update format string for the logging handler to include the recog XML filename
     logging.basicConfig(force=True, format=f"{sys.argv[1]}: {BASE_LOG_FORMAT}")
 
-    update_cpes(sys.argv[1], cpe_vp_map, r7_vp_map, deprecated_map)
+    files = sys.stdin.readlines()
```

Comment:
I frequently used the script to update a single file after making changes rather than processing all of the XML files. What are your thoughts on getting the same effect by processing the first parameter with glob thus allowing single or multi-file processing?

Reply:
I agree. Unless I'm missing something you'd have to `echo "xml/myfile.xml | ` . It might be better to have a flag to indicate the file.